src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -882,25 +846,38 @@ end
 Get the flow variables of component k
 """
 @inline function get_component(k, u, equations::IdealMhdMultiIonEquations2D)
-  return SVector(u[(k-1)*5+4],
-                 u[(k-1)*5+5],
-                 u[(k-1)*5+6],
-                 u[(k-1)*5+7],
-                 u[(k-1)*5+8])
+  # The first 3 entries of u contain the magnetic field. The following entries contain the density, momentum (3 entries), and energy of each component.
+  return SVector(u[3 + (k - 1) * 5 + 1],
+                 u[3 + (k - 1) * 5 + 2],
+                 u[3 + (k - 1) * 5 + 3],
+                 u[3 + (k - 1) * 5 + 4],
+                 u[3 + (k - 1) * 5 + 5])
+end
+
+"""
+Set the flow variables of component k
+"""
+@inline function set_component!(u, k, u1, u2, u3, u4, u5, equations::IdealMhdMultiIonEquations2D)
+  # The first 3 entries of u contain the magnetic field. The following entries contain the density, momentum (3 entries), and energy of each component.
+  u[3 + (k - 1) * 5 + 1] = u1
+  u[3 + (k - 1) * 5 + 2] = u2
+  u[3 + (k - 1) * 5 + 3] = u3
+  u[3 + (k - 1) * 5 + 4] = u4
+  u[3 + (k - 1) * 5 + 5] = u5
 end
 
 @inline function density_product(u, equations::IdealMhdMultiIonEquations2D)
   prod = one(u[1])
   for k in eachcomponent(equations)
-    prod *= u[(k-1)*5+4]
+    prod *= u[3 + (k - 1) * 5 + 1]
   end
   return prod
 end
 
 @inline function density(u, equations::IdealMhdMultiIonEquations2D)
   rho = 0
```

Comment:
I just noticed that here's another potential type instability and that this should be
```suggestion
  rho = zero(eltype(u))
```

Reply:
Fixed in 76325d0